nexus/operation-ids.md

```diff
@@ -0,0 +1,172 @@
+      - [Mapping from client to handler identifiers](#mapping-from-client-to-handler-identifiers)
+  - [Handler generated identifiers](#handler-generated-identifiers)
+    - [Cancel before started](#cancel-before-started)
+  - [Comparing the approaches](#comparing-the-approaches)
+    - [Handler Id qualification](#handler-id-qualification)
+    - [Ids as inputs](#ids-as-inputs)
+    - [Caller friendliness](#caller-friendliness)
+    - [Data sensitivity](#data-sensitivity)
+- [Proposal](#proposal)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Nexus RPCs - short recap
+
+At the fundamental level, the Nexus protocol is a synchronous RPC protocol. Asynchronous operations are modelled on top
+of a set of pre-defined RPCs.
+
+A Nexus **caller** calls a **handler**. The handler may respond inline or return a reference for a future, asynchronous
+operation. These asynchronous operations can be cancelled and checked for their outcomes and current state. The caller
+can also specify a callback URL, which the handler uses to asynchronously deliver the result of an operation's when it
+is ready.
+
+## Operation identifiers
+
+Since an asynchronous operation is modelled on top of synchronous RPCs, it must be addressable (identifiable) by both
+the caller and the handler. There are two approaches for generating identifiers, caller side and handler side.
+
+### Caller generated identifiers
+
+With this approach, the caller provides an Id to each of the pre-defined RPCs.
+
+The handler _should_ qualify an identifier when dealing with untrusted sources and multi-tenant applications.
+
+The following set of RPCs would be needed to model asynchronous operations using client generated Ids:
+
+- `Start(operation, id, input, callbackURL)` - handler responds inline or indicates that an async operation was started
+- `Cancel(operation, id)` - cancel operation by Id
+- `GetResult(operation, id)` - get operation result by Id
+- `Describe(operation, id)` - get operation status by Id
+- `DeliverResult(token)` - deliver a result of an async operation using a token in the Start request's callback URL
+
+Example invocations:
+
+```js
+Start(
+  'charge',
+  transactionId,
+  { customerId: 'someCustomerId', amount: 5000 },
+  'nexus://caller.example.com/callback?token=' + token
+);
+Cancel('charge', transactionId);
+```
+
+#### Handler addressability
+
+In some cases, the handler might require more than just a single identifier for a given operation, e.g. when an
+operation is backed by a workflow update, which requires both a workflow Id and an update Id. In order to avoid
+maintaining a handler side persistent mapping between caller and handler identifiers, the handler may choose to "leak"
+this detail to the caller, which would need to concatinate multiple identifiers for a given call. The example above
+becomes:
+
+```js
+Start(
+  'charge',
+  customerId + '/' + transactionId,
```

Comment:
I don't really follow how this is distinguished from the previous example. Customer ID moves from being a field in the input, to part of this ID - and I guess the point there is so it can be passed to things like `Cancel` which don't accept input.

IMO if we did end up doing something along these lines, I think you'd probably want the `path` argument to become a properly defined type rather than doing some concatenation. That way operations can explicitly define one-or-more typed IDs they care about.

Reply:
That's definitely something that I've considered but decided to keep it flat for simplicity (modeled after URL paths).
I'm not ruling it out though, agree that forcing the handler and client to encode and decode input to/from strings is suboptimal.

Reply:
I'm changing the proposal to accept optional caller generated IDs, those IDs may be random idempotency keys, addressable paths, etc...

I think the addressability concern in caller generated IDs is minor, if we provide a persistent mapping from caller ID to handler operation, it becomes a non issue at the cost of an extra persistence hit.
Handler for low latency operations that require complex identifiers for addressability could choose to disallow caller IDs or require data to be encoded in the ID.